src/backend/d3d12/CommandBufferD3D12.cpp

```diff
@@ -0,0 +1,183 @@
+// Copyright 2017 The NXT Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "CommandBufferD3D12.h"
+
+#include "common/Commands.h"
+#include "D3D12Backend.h"
+#include "PipelineD3D12.h"
+#include "PipelineLayoutD3D12.h"
+
+using Microsoft::WRL::ComPtr;
+
+namespace backend {
+namespace d3d12 {
+
+      CommandBuffer::CommandBuffer(Device* device, CommandBufferBuilder* builder)
+          : CommandBufferBase(builder), device(device), commands(builder->AcquireCommands()) {
+      }
+
+      CommandBuffer::~CommandBuffer() {
+          FreeCommands(&commands);
+      }
+
+      void CommandBuffer::FillCommands(ComPtr<ID3D12GraphicsCommandList> commandList) {
+          Command type;
+          Pipeline* lastPipeline = nullptr;
+
+          BeginRenderPassCmd* currentRenderPass = nullptr;
+
+          while(commands.NextCommandId(&type)) {
+              switch (type) {
+                  case Command::AdvanceSubpass:
+                      {
+                      }
+                      break;
+
+                  case Command::BeginRenderPass:
+                      {
+                          currentRenderPass = commands.NextCommand<BeginRenderPassCmd>();
```

Comment:
Not sure what you mean by global scope. Just so I'm clear what you mean, you mean just pull out `renderPass` and `framebuffer` at the start of a render pass to the top of the while loop?